annotation.go

```diff
@@ -101,26 +124,99 @@ type AnnotationEvent struct {
+	ID        platform.ID        `db:"id"`        // ID is the annotation's id.
+	OrgID     platform.ID        `db:"org_id"`    // OrgID is the annotations's owning organization.
+	StreamID  platform.ID        `db:"stream_id"` // StreamID is the id of a stream.
+	StreamTag string             `db:"name"`      // StreamTag is the name of a stream (when selecting with join of streams).
+	Summary   string             `db:"summary"`   // Summary is the summary of the annotated event.
+	Message   string             `db:"message"`   // Message is a longer description of the annotated event.
+	Stickers  AnnotationStickers `db:"stickers"`  // Stickers are additional labels to group annotations by.
+	Duration  string             `db:"duration"`  // Duration is the time range (with zone) of an annotated event.
+	Lower     string             `db:"lower"`     // Lower is the time an annotated event begins.
+	Upper     string             `db:"upper"`     // Upper is the time an annotated event ends.
+}
+
+// ToCreate is a utility method for converting a StoredAnnotation to an AnnotationCreate type
+func (s StoredAnnotation) ToCreate() (*AnnotationCreate, error) {
+	et, err := time.Parse(time.RFC3339Nano, s.Upper)
+	if err != nil {
+		return nil, err
+	}
+
+	st, err := time.Parse(time.RFC3339Nano, s.Lower)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AnnotationCreate{
+		StreamTag: s.StreamTag,
+		Summary:   s.Summary,
+		Message:   s.Message,
+		Stickers:  s.Stickers,
+		EndTime:   &et,
+		StartTime: &st,
+	}, nil
+}
+
+// ToEvent is a utility method for converting a StoredAnnotation to an AnnotationEvent type
+func (s StoredAnnotation) ToEvent() (*AnnotationEvent, error) {
+	c, err := s.ToCreate()
+	if err != nil {
+		return nil, err
+	}
+
+	return &AnnotationEvent{
+		ID:               s.ID,
+		AnnotationCreate: *c,
+	}, nil
+}
+
+type AnnotationStickers map[string]string
+
+// Value implements the database/sql Valuer interface for adding AnnotationStickers to the database
+// Stickers are stored in the database as a slice of strings like "[key=val]"
+// They are encoded into a JSON string for storing into the database, and the JSON sqlite extension is
+// able to manipulate them like an object.
+func (a AnnotationStickers) Value() (driver.Value, error) {
+	stickSlice := make([]string, 0, len(a))
+
+	for k, v := range a {
+		stickSlice = append(stickSlice, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	sticks, err := json.Marshal(stickSlice)
+	if err != nil {
+		return nil, err
+	}
+
+	return string(sticks), nil
+}
+
+// Scan implements the database/sql Scanner interface for retrieving AnnotationStickers from the database
+// The string is decoded into a slice of strings, which are then converted back into a map
+func (a *AnnotationStickers) Scan(value interface{}) error {
+	var stickSlice []string
+	if err := json.NewDecoder(strings.NewReader(value.(string))).Decode(&stickSlice); err != nil {
```

Comment:
Should we have any sort of guard on the type-cast to `string`? Not sure how many guardrails the sqlite/sql libraries are providing for us.

Reply:
I added some error handling logic to the type-cast! It's hard to imagine how it would ever come out of the database as something other than a string since that's what the column is, but it never hurts to double check.